### ISSUE TITLE
Cuda memory error when perform evalution

### ISSUE THREAD

**Issue by John1231983** [ChunhuanLin/deform_conv_pytorch#1]
I am running your code in python with TitanX. After completed training, I got below error. I am not sure that this error comes from GPU issue or not. How can I fix it? Thanks
```
Train Epoch: 1 [59520/60000 (99%)]	Loss: 0.255778
Train Epoch: 1 [59840/60000 (100%)]	Loss: 0.313443
Spends 9.129980778694152s for each training epoch
demo.py:173: UserWarning: volatile was removed and now has no effect. Use `with torch.no_grad():` instead.
  data, target = Variable(data, volatile=True), Variable(target)
THCudaCheck FAIL file=/home/john/pytorch/aten/src/THC/generic/THCStorage.cu line=58 error=2 : out of memory
Traceback (most recent call last):
  File "demo.py", line 190, in <module>
    test()
  File "demo.py", line 174, in test
    output = model(data)
  File "/home/john/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 371, in __call__
    result = self.forward(*input, **kwargs)
  File "demo.py", line 82, in forward
    x = F.relu(self.conv4(x, offsets))
  File "/home/john/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 371, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/john/deform_conv_pytorch/deform_conv.py", line 59, in forward
    x_q_lt = self._get_x_q(x, q_lt, N)
  File "/home/john/deform_conv_pytorch/deform_conv.py", line 117, in _get_x_q
    x_offset = x.gather(dim=-1, index=index).contiguous().view(b, c, h, w, N)
RuntimeError: cuda runtime error (2) : out of memory at /home/john/pytorch/aten/src/THC/generic/THCStorage.cu:58

```

**Comment by forrestl111** [comment]
It seems that we use different versions of PyTorch. [Volatile](http://pytorch.org/docs/0.3.0/notes/autograd.html?highlight=volatile) is set in my code to save memory when perform evaluation, but it seems not work with the PyTorch version you used. You may refer this [post](https://discuss.pytorch.org/t/torch-no-grad/12296), and modify your code as suggested.

**Comment by John1231983** [comment]
Hi @ChunhuanLin : I have read it as your suggestion. I change the code
```
        with torch.no_grad():
            data, target = Variable(data, volatile=True), Variable(target)
```
However, it still got the memory error. Do you know how to fix it? I am using pytorch 0.4.0 build from source

**Comment by forrestl111** [comment]
In fact, you should change the code in this way:
```   
data, target = Variable(data, requires_grad=True), Variable(target)   
with torch.no_grad():  
    output = model(data) 
```  
Refer the [official docs](http://pytorch.org/docs/master/autograd.html) for more information.

**Comment by John1231983** [comment]
Sorry, this issue still not solved 
```
THCudaCheck FAIL file=/home/john/pytorch/aten/src/THC/generic/THCStorage.cu line=58 error=2 : out of memory
Traceback (most recent call last):
  File "demo.py", line 194, in <module>
    test()
  File "demo.py", line 176, in test
    output = model(data)
  File "/home/john/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 371, in __call__
    result = self.forward(*input, **kwargs)
  File "demo.py", line 82, in forward
    x = F.relu(self.conv4(x, offsets))
  File "/home/john/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 371, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/john/deform_conv_pytorch/deform_conv.py", line 59, in forward
    x_q_lt = self._get_x_q(x, q_lt, N)
  File "/home/john/deform_conv_pytorch/deform_conv.py", line 117, in _get_x_q
    x_offset = x.gather(dim=-1, index=index).contiguous().view(b, c, h, w, N)
RuntimeError: cuda runtime error (2) : out of memory at /home/john/pytorch/aten/src/THC/generic/THCStorage.cu:58
```

**Comment by forrestl111** [comment]
I have reduced the batch size of testset, everything shall work well now.

**Comment by John1231983** [comment]
Thanks for your correction. But I got a new error about testing. I am using pytorch 0.4 builds from source
```
demo.py:179: UserWarning: invalid index of a 0-dim tensor. This will be an error in PyTorch 0.5. Use tensor.item() to convert a 0-dim tensor to a Python number
  test_loss += F.nll_loss(output, target, size_average=False).data[0] # sum up batch loss
Traceback (most recent call last):
  File "demo.py", line 194, in <module>
    test()
  File "demo.py", line 186, in test
    100. * correct / len(test_loader.dataset)))
RuntimeError: value cannot be converted to type uint8_t without overflow: 10000

```

**Comment by forrestl111** [comment]
Sorry, this repo is implemented by [PyTorch v-0.3.0](http://pytorch.org/docs/0.3.0/), and version 0.4 is not the stable version yet. So maybe it's a better choice to use [stable version](http://pytorch.org/docs/0.3.0/) of PyTorch for you. By the way, I would like to update the code when version 0.4 comes to be the stable version.
